### ISSUE TITLE
Ability to generate Dynamic Tests in an Extension

### ISSUE THREAD

**Issue by vmassol** [junit-team/junit5#1444]
Hi guys, I currently have a JUnit4 custom Suite https://github.com/xwiki/xwiki-rendering/blob/2e7af2e60fd6661ec48142324cd3fe10411fe0bc/xwiki-rendering-test/src/main/java/org/xwiki/rendering/test/integration/RenderingTestSuite.java#L78 which automatically registers new tests to the junit runner and which allows to write very simple tests:

```
@RunWith(RenderingTestSuite.class)
public class IntegrationTests
{
}
```

The custom suite reads `*.test` files found in the CL's classpath and generate dynamic tests from them.

I'd like to move this to JUnit5 but I'm a bit stuck ATM since it seems the best approach would be to use some test factory or parametrized test but this would require each of my tests to declare a test method with something like:

```
class IntegrationTests
{
   @TestFactory
   Stream<DynamicTest> renderingTests()
   {
           return RenderingTestSuite.generateStreamOfDynamicTests();
   }
}
```

I'd really like to be able that my test keep being as simple as the following (and this without duplication):

```
@ExtendWith(RenderingExtension.class)
public class IntegrationTests
{
}
```

PS: I've also asked the question on Gitter: https://gitter.im/junit-team/junit5?at=5b142454c712f5612555bfcc (which contains other details).

**Comment by sbrannen** [comment]
Tentatively assigned to the 3.x backlog for _team discussion_

**Comment by vmassol** [comment]
Thanks @sbrannen

**Comment by sbrannen** [comment]
Closely related to #371.

**Comment by kriegfrj** [comment]
Any movement on this one? I have a use case that I think I can *almost* implement using standard extensions points, but not quite.

You guys are no doubt familiar with soft assertions in AssertJ - it collates the test failures into a single `MultipleFailuresError` and throws that instead so you have perform several assertions per test and not have the first failure prevent the rest of the assertions running.

A problem we are having is dealing with the verbosity of the resulting error message, in particular if we add the stack traces to it for easy IDE navigation.

I had a crazy idea - instead of collating them into a single error and reporting them all against a single test node in the test hierarchy, each of the assertions would generate a `DynamicTest` node and store the success/failure against it. This is also inspired in part by `Specsy` which has a similar philosophy of generating multiple nodes, but it allows you to leverage the richness of AssertJ's fluent assertions to do so.

There are two main ways I could implement it. One of them is to come up with a soft proxy generator that produces a stream or collection of `DynamicTest` nodes as a return value. This could then be used inside of a `@TestFactory` method. The disadvantage of this is that it would require major refactoring of existing tests using soft proxies if the developer wishes to leverage it.

The alternative is to have the smarts in an extension's `afterEach()` callback to transform the collected assertions into `DynamicTest` nodes. This would make for an extremely simply migration path - the developer simply needs to change their `@ExtendWith` declaration to use the new extension, and it would work without needing to change any of the individual tests. But this is not possible with the current API.

My proposal is quite simple: add a `registerTest(DynamicNode)` method to `ExtensionContext`. This method will allow any extension to register a new test node as a child of the currently-executing node. It would solve my use case as I could call `registerTest()` from within my extensions `afterEach()` callback.

Thoughts?

**Comment by kriegfrj** [comment]
> ## Overview
> I currently have a JUnit 4 based [custom `Suite`](https://github.com/xwiki/xwiki-rendering/blob/2e7af2e60fd6661ec48142324cd3fe10411fe0bc/xwiki-rendering-test/src/main/java/org/xwiki/rendering/test/integration/RenderingTestSuite.java#L78) which automatically registers new tests to the JUnit runner and which allows to write very simple tests:
> 
> ```java
> @RunWith(RenderingTestSuite.class)
> public class IntegrationTests {
> }
> ```
> 
> The custom suite reads `*.test` files found in the CL's classpath and generates dynamic tests from them.
> 
> I'd like to move this to JUnit 5, but I'm a bit stuck ATM since it seems the best approach would be to use some test factory or parameterized test, but this would require each of my tests to declare a test method with something like:
> 
> ```java
> class IntegrationTests {
> 
>    @TestFactory
>    Stream<DynamicTest> renderingTests() {
>            return RenderingTestSuite.generateStreamOfDynamicTests();
>    }
> }
> ```
> 
> I'd really like my tests to remain as simple as the following (and this without duplication):
> 
> ```java
> @ExtendWith(RenderingExtension.class)
> public class IntegrationTests {
> }
> ```
> 
> PS: I've also asked the question on [Gitter](https://gitter.im/junit-team/junit5?at=5b142454c712f5612555bfcc) (which contains additional details).
> 

@vmassol, it also occurred to me that your problem could be solved with the existing API, as follows:

```java
public interface RenderingTests {
    @TestFactory
    default Stream<DynamicTest> renderingTests() {
            return RenderingTestSuite.generateStreamOfDynamicTests();
    }
}

class IntegrationTests implements RenderingTests {
    ...
}
```

Hope that helps (although I see it was a long time ago... you've probably either solved it or moved on!)

I still would like to push for the API to allow extensions to register dynamic tests, as I don't think my use case has a workaround using the existing API.

**Comment by vmassol** [comment]
> Hope that helps (although I see it was a long time ago... you've probably either solved it or moved on!)

@kriegfrj Yes that could be a workaround. I still consider that a workaround and I'd prefer to have a way to register dynamic tests in a JUnit Extension. That would be cleaner and nicer.

Thanks for the idea!

**Comment by kriegfrj** [comment]
I'm still interested in this one. Haven't given up on the long-term AssertJ plan.

Pinging @joel-costigliola, @scordio from the AssertJ team. I will raise an issue there too and link it to this one.

**Comment by vlsi** [comment]
I wonder if generating tests during test execution will be compatible with "restart test from IDE", "run specific test from command line", "filter tests by tag".

**Comment by jbduncan** [comment]
JUnit 5 team, it looks like the stale bot forgot to remove the "stale" label for this issue and subsequently closed it too early. Would you mind re-opening it?

**Comment by mpkorstanje** [comment]
> The custom suite reads *.test files found in the CL's classpath and generates dynamic tests from them.
> I'd like to move this to JUnit 5, but I'm a bit stuck ATM since it seems the best approach would be to use some test factory or parameterized test, but this would require each of my tests to declare a test method with something like:

This doesn't look like something that should be done by JUnit Jupiter, but rather a custom `TestEngine` implementation. To work around the limited support for custom test engines in Maven and Gradle you can then use the Suite engine to call the custom test engine.

```java
@Suite
@IncludeEngines("rendering-suite")
@SelectPackages("com.example.app")
class IntegrationTests {
}
```

**Comment by mpkorstanje** [comment]
> My proposal is quite simple: add a registerTest(DynamicNode) method to ExtensionContext. This method will allow any extension to register a new test node as a child of the currently-executing node. It would solve my use case as I could call registerTest() from within my extensions afterEach() callback.
>
> Thoughts?

@kriegfrj the different assertions are more like steps in a test rather then independent tests. Trying to represent them as tests seems like a conceptual mismatch. Your proposal is also backed by a very different use case, I don't think it is well served by being combined with this issue.

edit: I also see that Specsy has a JUnit Engine implementation. Have you considered using Specsy?

**Comment by sormuras** [comment]
"more like steps in a test" reminds me of:
- #703 and
- #1821

**Comment by marcphilipp** [comment]
**Team decision:** `@TestFactory` is sufficiently flexible for one-off scenarios like in the issue description. For more frequent occurrences, we recommend writing a custom `TestEngine`.

**Comment by kriegfrj** [comment]
Sorry for the late reply! I see the decision was made to close this - I don't expect this response to change that decision, but thought I would respond to the points that were made.

> > My proposal is quite simple: add a registerTest(DynamicNode) method to ExtensionContext. This method will allow any extension to register a new test node as a child of the currently-executing node. It would solve my use case as I could call registerTest() from within my extensions afterEach() callback.
> > Thoughts?
> 
> [@kriegfrj](https://github.com/kriegfrj) the different assertions are more like steps in a test rather then independent tests. Trying to represent them as tests seems like a conceptual mismatch.

I think this is because we are conceptually constrained by the way JUnit was originally designed, using AssertionError which gave rise to the "one method, one assertion" approach. This is unique to JUnit and friends. I've had quite a bit of experience with Test::More and Test::Class::Moose in the Perl ecosystem and this method<->assertion equivalence is not present.

Take a step back (pun intended) to the pure conceptual level: to test a system, we need to 1. prepare its inputs, 2. run it, and 3. check the actual output(s) against the expected output(s). Wash, rinse, repeat for every different aspect of the system that we want to check.

Step three requires you to check each actual against each expected. Each of these could reasonably be conceived as a "test". You could test each output in its own method, but that would involve the runtime overhead of re-running 1 & 2 for each output that you want to check. Where step 1 & 2 are time consuming, it is more efficient to do them once and then test all outputs. This is also the reasoning behind features like `@BeforeAll`, to try and factor some of that common code out to prevent needing to re-run it.

> Your proposal is also backed by a very different use case, I don't think it is well served by being combined with this issue.
> 

It is true that my use case is different, but I don't see that that matters too much. It's the mark of a powerful and flexible feature if it can support multiple different use cases.

My use case is also not well served by either a `@TestFactory` or a custom `TestEngine` - either will require a significant rewrite of an existing body of tests. I suppose I could extend `JupiterEngine` to add the functionality that I need rather than re-writing a custom `TestEngine` from scratch.

> edit: I also see that Specsy has a JUnit Engine implementation. Have you considered using Specsy?

I referenced Specsy in the original proposal. From memory, it didn't play well with OSGi out-of-the-box. That could perhaps be overcome, but more significantly using it would require a complete rewrite of all the (extensive) body of tests that I already have that are using Soft Assertions and Jupiter TestEngine. Aside from the time investment required to do this, there is always a significant risk when refactoring tests that one might inadvertently break them in a way that they always return a "pass" even when the feature they are supposedly testing has been broken.

**Comment by mpkorstanje** [comment]
> My use case is also not well served by either a @TestFactory or a custom TestEngine - either will require a significant rewrite of an existing body of tests. 

Sorry, let me rephrase that. I think your use case would be better served by creating a separate issue. By posing a problem, without going into a solution, there is more room to consider alternative angles.
